openlibrary/templates/type/edition/view.html

```diff
@@ -24,39 +24,49 @@
   $# We are on an editions page: an edition has been explicitly selected
   $ edition = page
 
+$ selected_provider = None
+$ selected_id = None
+$if query_param('edition'):
+  $if ':' in query_param('edition'):
+    $ [selected_provider, selected_id] = query_param('edition').split(':')
+  $else:
+    $ [selected_provider, selected_id] = ['ia', query_param('edition')]
+
 $ component_times['get_sorted_editions'] = time()
 
 $# Fetch a work's editions
 $# Used for loading the editions table
 $# Book availability of editions injected by bulk get_availability API
 $ edition_count = work.edition_count if work and work.edition_count else 1
-$ ebooks_only = query_param('mode') == 'ebooks' or (query_param('mode') != 'all' and edition_count > 10)
+$ ebooks_only = (query_param('mode') == 'ebooks') or (query_param('mode') != 'all' and edition_count > 10)
 
 $# For performance reasons, limit to 10 ebooks
 $# Tradeoff: limits our ability to select best edition
-$ editions_limit = None if not query_param('mode') else 10
+$ editions_limit = 10 if not query_param('mode') else None
+$# key ensures the current edition we're on or requested edition are fetched by get_sorted_editions
+$ keys = (edition and [edition.key]) or (selected_provider=="ia" and selected_id and [work.get_olid_by_ocaid(selected_id)])
 $if ebooks_only:
-  $ editions = work.get_sorted_editions(ebooks_only=ebooks_only, limit=editions_limit, keys=edition and [edition.key])
-
+  $ editions = work.get_sorted_editions(ebooks_only=ebooks_only, limit=editions_limit, keys=keys)
 $if not editions:
   $# use ebooks_only to determine whether to lazyload editions table
   $ ebooks_only = False
-  $ editions = work.get_sorted_editions(limit=editions_limit, covers_only=True)
+  $# if there's a limit, attempt to choose books w/ covers
+  $ editions = work.get_sorted_editions(limit=editions_limit, covers_only=True, keys=keys)
+$if not editions:
+  $# if there weren't any editions with covers, just fetch what we have (edge case)
+  $ editions = work.get_sorted_editions(limit=editions_limit, keys=keys)
 $ availabilities = {e.availability.get('identifier'): e.availability for e in editions}
 $ component_times['get_sorted_editions'] = time() - component_times['get_sorted_editions']
 
 $# This collects which books are previewable in any manner
 $ previews = [e for e in editions if e.get('ocaid')]
 
 $# Choose an edition to render
-$if not edition:
+$if editions and not edition:
   $# We're presumably on a work url
   $# edition selection strategy (e.g. language, by id, first w/ ocaid)
-  $if query_param('edition'):
-    $if ':' in query_param('edition'):
-      $ [selected_provider, selected_id] = query_param('edition').split(':')
-    $else:
-      $ [selected_provider, selected_id] = ['ia', query_param('edition')]
+  $ edition = editions[0]
```

Comment:
Is that true? Even in the case where an edition id is specified in the url but it's a wrong id? I thought this was one of the problems we were trying to solve. Otherwise, we were/are hitting a case where we get to the line `edition = edition or page` and the result was `edition` was getting assigned the `work`.

Reply:
Read the if/else clause; they both set `edition` to something. This line is never going to get used.